# Zen of python

In [4]:
import this
import pandas
import pandas as pd

# The Python Programming Language: Types and Sequences

Use `type` to return the object's type.

In [13]:
df = pd.read_csv('class_mpg.csv', index_col=0)

In [14]:
df.head(1)

manufacturer model  displ  year  cyl     trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4  auto(l5)   f   18   29  p  compact

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [12]:
type(df['class'][0])

str

In [26]:
len(df[df['class']!='compact'])

187

In [29]:
df.rename(columns = {'class':"mode"})

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

        mode  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

# The Python Programming Language: Functions

`add_numbers` is a function that takes two numbers and adds them together.

In [36]:
def add_numbers(x, y, z):
    return x+y+z

In [33]:
add_numbers(2,3)

5

Try calling the function using strings

In [34]:
add_numbers('a','b')

'ab'

Now using a 3rd parameter

In [37]:
add_numbers(1,2,3)

6

Support an additional parameter

In [ ]:
def add_numbers(x, y):
    pass

`add_numbers` updated to take an optional flag parameter. Using `print` allows printing of multiple expressions within a single cell.

In [38]:
def add_numbers(x, y, z= None, flag=False):
    if flag:
        print('true')
    if not z:
        return x + y

    else:

        return x+y+z

Assign function `add_numbers` to variable `a`.

In [ ]:
def add_numbers(x,y):
    pass

# Tuples & Lists


## Lists
Lists are a mutable data structure.

In [39]:
a = [2, 3, 4, 'Max']

Use `append` to append an object to a list.

In [40]:
a.append('tom')

This is an example of how to loop through each item in the list.

In [41]:
a[0]

2

**len()** Für die Anzahl der Einträge

In [46]:
for i in range(len(a)):
    print(a[i])

2
3
4
Max
tom


Or using the indexing operator:

In [45]:
len(a)

5

Use `+` to concatenate lists.

Use `*` to repeat lists.

Use the `in` operator to check if something is inside a list.

Ab der dritten Stelle bis zum Ende

Von 2ter bis 5ter stelle

Jedes Element in sliced mal 2 rechnen. Ergebnis = [6, 8, 10, 12]

Python way = List comprehension

## Tuples

## Strings
Now let's look at strings. Use bracket notation to slice a string.

This will return the last element of the string.

This is a slice from the beginning of the string and stopping before the 3rd element.

And this is a slice starting from the 4th element of the string and going all the way to the end.

**Concatenate**

`in` operator also with strings

`split` returns a list of all the words in a string, or a list split on a specific character.

Make sure you convert objects to strings before concatenating.

Python has a built in method for convenient string formatting.

In [ ]:
sales_record = {
'price': 3.24,
'num_items': 4,
'person': 'Chris'}

sales_statement = '{} bought {} item(s) at a price of {} each for a total of {}'

print(sales_statement.format(sales_record['person'],
                             sales_record['num_items'],
                             sales_record['price'],
                             sales_record['num_items']*sales_record['price']
                             ))


## Dictionary (aka hashtable)

Iterate over all of the keys:

Iterate over all of the values:

Iterate over all of the items in the list:

## Unpacking

You can unpack a sequence into different variables:

In [ ]:
def values():
    return 1,2

Make sure the number of values you are unpacking matches the number of variables being assigned.

# Reading and Writing CSV files

Let's import our datafile mpg.csv, which contains fuel economy data for 234 cars.

* mpg : miles per gallon
* class : car classification
* cty : city mpg
* cyl : # of cylinders
* displ : engine displacement in liters
* drv : f = front-wheel drive, r = rear wheel drive, 4 = 4wd
* fl : fuel (e = ethanol E85, d = diesel, r = regular, p = premium, c = CNG)
* hwy : highway mpg
* manufacturer : automobile manufacturer
* model : model of car
* trans : type of transmission
* year : model year

In [ ]:
!cat class_mpg.csv

In [ ]:
import csv

In [ ]:
with open('class_mpg.csv') as csvfile:
    mpg = list(csv.DictReader(csvfile))

`csv.Dictreader` has read in each row of our csv file as a dictionary. `len` shows that our list is comprised of 234 dictionaries.

`keys` gives us the column names of our csv.

This is how to find the average cty fuel economy across all cars. All values in the dictionaries are strings, so we need to convert to float.

Similarly this is how to find the average hwy fuel economy across all cars.


Use `set` to return the unique values for the number of cylinders the cars in our dataset have.

Here's a more complex example where we are grouping the cars by number of cylinder, and finding the average cty mpg for each group.

In [ ]:
CtyMpgByCyl = []

for c in cylinders: # iterate over all the cylinder levels
    summpg = 0
    cyltypecount = 0
    for d in mpg: # iterate over all dictionaries
        if d['cyl'] == c: # if the cylinder level type matches,
            summpg += float(d['cty']) # add the cty mpg
            cyltypecount += 1 # increment the count
    CtyMpgByCyl.append((c, summpg / cyltypecount)) # append the tuple ('cylinder', 'avg mpg')

#CtyMpgByCyl.sort(key=lambda x: x[0])
CtyMpgByCyl

In [ ]:
vehicleclass = set([row['class'] for row in mpg])
vehicleclass

And here's an example of how to find the average hwy mpg for each class of vehicle in our dataset.

In [ ]:
HwyMpgByClass = []

for t in vehicleclass: # iterate over all the vehicle classes
    summpg = 0
    vclasscount = 0
    for d in mpg: # iterate over all dictionaries
        if d['class'] == t: # if the cylinder amount type matches,
            summpg += float(d['hwy']) # add the hwy mpg
            vclasscount += 1 # increment the count
    HwyMpgByClass.append((t, summpg / vclasscount)) # append the tuple ('class', 'avg mpg')

HwyMpgByClass.sort(key=lambda x: x[1])
HwyMpgByClass

# The Python Programming Language: Dates and Times

In [ ]:
import datetime as dt
import time as tm

`time` returns the current time in seconds since the Epoch. (January 1st, 1970)

Convert the timestamp to datetime.

Handy datetime attributes:

`timedelta` is a duration expressing the difference between two dates.

`date.today` returns the current local date.


# The Python Programming Language: Objects and map()

An example of a class in python:

In [ ]:
class Person:
    department = "School"
    
    def set_name(self, new_name):
        self.name = new_name
    def say_name(self):
        print(self.name)

Here's an example of mapping the `min` function between two lists.

Now let's iterate through the map object to see the values.

# The Python Programming Language: Lambda and List Comprehensions

Here's an example of lambda that takes in three parameters and adds the first two.

Let's iterate from 0 to 999 and return the even numbers.

Now the same thing but with list comprehension.

# The Python Programming Language: Numerical Python (NumPy)

In [ ]:
import numpy as np

## Creating Arrays

Create a list and convert it to a numpy array

Or just pass in a list directly

Pass in a list of lists to create a multidimensional array.

Use the shape method to find the dimensions of the array. (rows, columns)

`arange` returns evenly spaced values within a given interval.

`reshape` returns an array with the same data with a new shape.

`linspace` returns evenly spaced numbers over a specified interval.

`resize` changes the shape and size of array in-place.

`ones` returns a new array of given shape and type, filled with ones.

`zeros` returns a new array of given shape and type, filled with zeros.

`eye` returns a 2-D array with ones on the diagonal and zeros elsewhere.

`diag` extracts a diagonal or constructs a diagonal array.

Create an array using repeating list (or see `np.tile`)

Repeat elements of an array using `repeat`.

#### Combining Arrays

Use `vstack` to stack arrays in sequence vertically (row wise).

Use `hstack` to stack arrays in sequence horizontally (column wise).

## Operations

Use `+`, `-`, `*`, `/` and `**` to perform element wise addition, subtraction, multiplication, division and power.

**Dot Product:**  

$ \begin{bmatrix}x_1 \ x_2 \ x_3\end{bmatrix}
\cdot
\begin{bmatrix}y_1 \\ y_2 \\ y_3\end{bmatrix}
= x_1 y_1 + x_2 y_2 + x_3 y_3$

Let's look at transposing arrays. Transposing permutes the dimensions of the array.

The shape of array `z` is `(2,3)` before transposing.

Use `.T` to get the transpose.


The number of rows has swapped with the number of columns.

Use `.dtype` to see the data type of the elements in the array.

Use `.astype` to cast to a specific type.

## Math Functions

Numpy has many built in math functions that can be performed on arrays.

Sum

Max

Min

Mean

Std

`argmax` and `argmin` return the index of the maximum and minimum values in the array.

argmax()

a.argmin()

## Indexing / Slicing

Use bracket notation to get the value at a specific index. Remember that indexing starts at 0.

Use `:` to indicate a range. `array[start:stop]`


Leaving `start` or `stop` empty will default to the beginning/end of the array.

Use negatives to count from the back.

A second `:` can be used to indicate step-size. `array[start:stop:stepsize]`

Here we are starting 5th element from the end, and counting backwards by 2 until the beginning of the array is reached.

Let's look at a multidimensional array.

Use bracket notation to slice: `array[row, column]`

And use : to select a range of rows or columns

Here we are selecting all the rows up to (and not including) row 2, and all the columns up to (and not including) the last column.

This is a slice of the last row, and only every other element.

We can also perform conditional indexing. Here we are selecting values from the array that are greater than 30. (Also see `np.where`)

Here we are assigning all values in the array that are greater than 30 to the value of 30.

In [ ]:
r[r>30] = 100

In [ ]:
r

## Copying Data

Be careful with copying and modifying arrays in NumPy!


`r2` is a slice of `r`

In [ ]:
# r2 = 0

Set this slice's values to zero ([:] selects the entire array)

`r` has also been changed!

To avoid this, use `r.copy` to create a copy that will not affect the original array

Now when r_copy is modified, r will not be changed.

### Iterating Over Arrays

Let's create a new 4 by 3 array of random numbers 0-9.


Iterate by row:


Iterate by index:

Iterate by row and index:

Use `zip` to iterate over multiple iterables.